apina-core/src/main/kotlin/fi/evident/apina/spring/JacksonTypeTranslator.kt

```diff
@@ -96,8 +96,15 @@ internal class JacksonTypeTranslator(private val settings: TranslationSettings,
                 ApiType.Dictionary(translateType(arguments[1], env))
             classes.isInstanceOf<Optional<*>>(baseType) && arguments.size == 1 ->
                 ApiType.Nullable(translateType(arguments[0], env))
-            else ->
-                translateType(baseType, env)
+            else -> {
+                arguments.filterIsInstance<JavaType.Basic>().forEach{
+                    val typeName = classNameForType(it)
+                    if (!settings.isImported(typeName) && !settings.isBlackBoxClass(it.name) && !hasJsonValueAnnotation(it) ) {
+                        addDefinitionIfNotExistsYet(it, typeName, env)
+                    }
+                }
+                return translateType(baseType, env)
```

Comment:
If I understand your intention correctly, I think that we could just call `translateType` for all arguments and ignore the results:

```kotlin
else -> {
    // make sure that classes are generated for all generic parameters
    for (argument in arguments)
        translateType(argument, env)

    return translateType(baseType, env)
}
```

Does this seem correct? At least your new tests pass with this change. This way no other changes are even needed and there's no duplication of logic.

Reply:
Yes, that is much better! :D